### ISSUE TITLE
`JsonCodecMaker.make` doesn't work if there's `@named` annotation using `inline val` definition.

### ISSUE THREAD

**Issue by doohochang** [plokhotnyuk/jsoniter-scala#1023]
```scala
sealed trait Foo
object Foo:
  @named(A.name)
  case class A(integer: Int) extends Foo
  object A:
    inline val name = "TypeA"

  @named(B.name)
  case class B(string: String) extends Foo
  object B:
    inline val name = "TypeB"

  given JsonValueCodec[Foo] = JsonCodecMaker.make[Foo](
    CodecMakerConfig
      .withDiscriminatorFieldName(Some("type"))
  )
```

In cases like the above, `JsonCodecMaker.make` macro returns compile error with message `Can't evaluate compile-time expression`.

`A.name` and `B.name` are determined at compile time, but it currently seems to support only string literals for the annotation value.

It would be nice to be able to use values that are determined at compile time, such as `inline val`.

**Comment by plokhotnyuk** [comment]
Hi, Dooho!

Yes, you are right, the Scala 3 macros cannot evaluate arbitrary expressions in a compile time.
Currently they fail with the following error message:
```
com.github.plokhotnyuk.jsoniter_scala.macros.CompileTimeEval$CompileTimeEvalException: Unsupported constant expression: Select(Select(This(Ident(Foo$)),A),name)
```
The `Can't evaluate compile-time expression` error could happen in case of the configuration parameter of the `make` macro call cannot be evaluated in a compile time.

Could you please share more context about our use case to find an acceptable solution?

**Comment by doohochang** [comment]
In our use-cases, we generally declare `inline val name = "..."`  to share the identical name to multiple annotations defined in different libraries. (ex. jsoniter, tapir, caliban)

I didn't look macro code in Jsoniter thoroughly, but I guess the string value provided into `@named` doesn't have to be a constant for deriving `JsonValueCodec` instance.

I think the macro code like below would work in more various situations.
```scala
val value: Expr[String] = annotationTerm match
  case Apply(_, term :: Nil) =>
    term.asExpr match
      case '{ $name: String } => Some(name)
      case _                  => None

  case other => report.errorAndAbort(s"Can't recognize annotation arguments ${other.show}")
```

**Comment by plokhotnyuk** [comment]
@doohochang Unfortunately your code sample does not work: `name` is expression, not a string instance here.

**Comment by doohochang** [comment]
To give a little more context, we are using some custom annotations in our project that are used for type class derivation.
The above code sample is being used in a macro function as shown below.
```scala
protected[macros] def getAnnotationValueForType[A: Type](using Quotes): Expr[String] =
  import quotes.reflect.*

  TypeRepr.of[A].typeSymbol.annotations
    .find(_.tpe =:= TypeRepr.of[my.custom.annotation])
    .flatMap {
      case Apply(_, term :: Nil) =>
        term.asExpr match
          case '{ $name: String } => Some(name)
          case _                  => None

      case other => report.errorAndAbort(s"Can't recognize annotation arguments ${other.show}")
    }
```
It extracts the value of the target annotation even if it's not a literal or constant, because the code matches all kind of expression typed with `String`.

But I haven't checked the codec derivation in Jsoniter, so I'm not sure if my code sample is applicable to Jsoniter as you said.

Thanks for the quick answer @plokhotnyuk ! Our experience with Jsoniter has been amazing.
I will check if I can contribute to this issue myself when possible.

**Comment by plokhotnyuk** [comment]
The challenge here that we need a string instance (not `Expr[String]`) and use it in a compile-time (not just pasting the expression to outgoing tree of the macro).

BTW, if the naming rule in your code is consistent as in the provided sample then you can use the following configuration for the `make` call:
`CodecMakerConfig.withAdtLeafClassNameMapper(x => "Type" + JsonCodecMaker.simpleClassName(x))`

**Comment by doohochang** [comment]
Ah I understood the current situation.
Are there any disadvantages to using expressions for codec instantiation? (maybe for performance?)

Like the sample you provided, we did some workaround by using `.withAdtLeafClassNameMapper` and `.withFieldNameMapper`, so the issue in our project has been resolved. Thanks for help.

**Comment by plokhotnyuk** [comment]
Yes, having a string for keys and values in the compile-time allow to use some speeding up techniques for serialization of ASCII only strings that do not require JSON escaping.

Also, they are validated in the compile-time to avoid clashing of renamed JSON keys or ADT type hint values.

I'm happy to hear that your issue is resolved now.

Please, keep trying and sending your feedback!
